unifi_protect_backup/unifi_protect_backup.py

```diff
@@ -289,6 +295,52 @@ async def rclone_purge_old():
                 logger.warn(f"stdout:\n{stdout.decode()}")
                 logger.warn(f"stderr:\n{stderr.decode()}")
 
+        # We need to catch websocket disconnect and trigger a reconnect.
+        @aiocron.crontab("* * * * *")
+        async def check_websocket_and_reconnect():
+            logger.debug("Checking the status of the websocket...")
+            if self._protect.check_ws():
+                logger.debug("OK")
+            else:
+                logger.warn("Lost connection to Unifi Protect.")
+
+                # Unsubscribe - not strictly necessary, but good practice.
+                self._unsub()
+
+                while True:
+                    logger.warn("Attempting reconnect...")
+
+                    try:
+                        # Start again from scratch. In principle if Unifi
+                        # Protect has not been restarted we should just be able
+                        # to call self._protect.update() to reconnect to the
+                        # websocket. However, if the server has been restarted a
+                        # call to self._protect.check_ws() returns true and some
+                        # seconds later pyunifiprotect detects the websocket as
+                        # disconnected again. Therefore, kill it all and try
+                        # again!
+                        self._protect = ProtectApiClient(
+                            self.address,
+                            self.port,
+                            self.username,
+                            self.password,
+                            verify_ssl=self.verify_ssl,
+                            subscribed_models={ModelType.EVENT},
+                        )
+                        # Start the pyunifiprotect connection by calling `update`
+                        await self._protect.update()
+                        if self._protect.check_ws():
+                            self._unsub = self._protect.subscribe_websocket(self._websocket_callback)
+                            break
+                        else:
+                            raise RuntimeError("Unable to establish connection to Unifi Protect")
```

Comment:
Does this need to be an exception or could it be a log?

Reply:
I've logged this as warning. Is that strong enough, or should it be an error?

Reply:
Hmm good question. I don't see errors used anywhere else. That might be something to review separately for the project as a whole. In my mind an error is something that means we cannot recover and must stop and since this is designed to never stop, we wont see many errors.